File: plugins/ghostwriter/agent/tools/lib/context_resolver.py
"""
context_resolver.py

Python equivalent of context-resolver.ts for Python tools.
This module wraps config_loader.py for backward compatibility
while supporting both the old context format and new config files.

Uses snake_case throughout for consistency.
"""

import json
import os
from typing import Any, Dict, Optional, TypedDict, Literal

from .config_loader import (
    ResolvedWritingConfig,
    ResolvedDetectionConfig,  # deprecated alias
    Publication,
    WritingStyle,
    load_config,
    load_preset,
    load_from_media_type,
    get_config_from_env,
    get_builtin_presets,
    get_media_types,
    serialize_config_for_env,
    MEDIA_TYPES,
)

# Legacy type definitions for backward compatibility
MediaType = Literal[
    "blog", "book", "technical", "academic", "casual", "student", "business", "unknown"
]
FormalityLevel = Literal["formal", "casual", "mixed"]


class StyleOptions(TypedDict, total=False):
    period_quote: Literal["american", "british", "logical"]
    oxford_comma: Literal["always", "never", "flexible"]


class ThresholdConfig(TypedDict, total=False):
    ai_threshold: Optional[float]
    human_baseline: float
    disabled: bool
    weight: float


def _is_media_type(value: str) -> bool:
    """Check if a string is a valid media type."""
    return value in MEDIA_TYPES


def get_config() -> ResolvedWritingConfig:
    """
    Get writing config from WRITING_CONFIG environment variable.

    Returns:
        ResolvedWritingConfig with at minimum { publication: { media: "unknown" } }
    """
    env_value = os.environ.get("WRITING_CONFIG", "")
    if env_value:
        try:
            return json.loads(env_value)
        except json.JSONDecodeError:
            pass

    return load_from_media_type("unknown")


# Deprecated alias
get_context = get_config


def resolve_config(input_value: Optional[str] = None) -> ResolvedWritingConfig:
    """
    Resolve config input to a ResolvedWritingConfig.

    Supports multiple input formats for backward compatibility:
    - None/empty: returns default config
    - Media type string: "blog" -> loads blog preset
    - Preset name: "technical-book" -> loads preset
    - JSON string: '{"media":"blog"}' -> legacy format, converts to config
    - File path: "./config.yml" -> loads full config

    Args:
        input_value: Config input string

    Returns:
        ResolvedWritingConfig
    """
    if input_value is None or input_value == "":
        return get_config()

    input_str = input_value.strip()

    # Check if it's a file path
    if (input_str.endswith(".yaml") or input_str.endswith(".yml") or
            input_str.startswith("./") or input_str.startswith("/")):
        import os.path
        if os.path.exists(input_str):
            return load_config(input_str)

    # Check if it's a media type
    if _is_media_type(input_str):
        return load_from_media_type(input_str)

    # Check if it's JSON (legacy format or new format)
    if input_str.startswith("{"):
        try:
            parsed = json.loads(input_str)

            # Check if it's the new ResolvedWritingConfig format
            if "rules" in parsed and "publication" in parsed:
                return parsed

            # Legacy format - extract media type
            media = parsed.get("media", "unknown")
            return load_from_media_type(media)
        except json.JSONDecodeError:
            pass

    # Check if it's a preset name
    try:
        return load_preset(input_str)
    except (ValueError, FileNotFoundError):
        pass

    # Default to unknown
    return load_from_media_type("unknown")


# Deprecated alias
resolve_context = resolve_config


def get_presets() -> Dict[str, ResolvedWritingConfig]:
    """
    Get all available presets as resolved configs.

    Returns:
        Dict mapping preset name to resolved config
    """
    result = {}
    for name in get_builtin_presets():
        try:
            result[name] = load_preset(name)
        except (ValueError, FileNotFoundError):
            pass
    return result


# Re-export for convenience
__all__ = [
    "ResolvedWritingConfig",
    "ResolvedDetectionConfig",  # deprecated alias
    "ThresholdConfig",
    "Publication",
    "WritingStyle",
    "get_config",
    "get_context",  # deprecated alias
    "resolve_config",
    "resolve_context",  # deprecated alias
    "get_presets",
    "get_media_types",
    "load_config",
    "load_preset",
    "load_from_media_type",
    "serialize_config_for_env",
]

File: plugins/ghostwriter/agent/tools/lib/context-resolver.ts
/**
 * context-resolver.ts
 *
 * Resolves writing configuration from various input formats.
 * Central utility for making tools config-aware.
 *
 * This module wraps config-loader.ts and provides the main API for
 * resolving configs from different input types (file paths, media types,
 * preset names, JSON strings).
 */

import { existsSync } from 'fs';
import type {
  ResolvedWritingConfig,
  MediaType,
  BaseRuleConfig,
  RuleCategory,
  Publication,
  WritingStyle,
  FormalityLevel,
} from '../config/schemas';
import { isMediaType, MEDIA_TYPES } from '../config/schemas';
import {
  loadConfig,
  loadPreset,
  loadFromMediaType,
  serializeConfigForEnv,
  deserializeConfigFromEnv,
  getBuiltinPresets,
} from './config-loader';

// Re-export types from config-loader
export type {
  ResolvedWritingConfig,
  MediaType,
  BaseRuleConfig,
  RuleCategory,
  Publication,
  WritingStyle,
};

/** @deprecated Use ResolvedWritingConfig instead */
export type ResolvedDetectionConfig = ResolvedWritingConfig;

// Cache for loaded config
let cachedConfig: ResolvedWritingConfig | null = null;
let cachedConfigPath: string | null = null;

/**
 * Clear the cached config (useful for testing).
 *
 * When the config is loaded from a file path, it's cached to avoid redundant
 * file I/O. Call this function to force a fresh reload on next resolution.
 *
 * @example
 * ```ts
 * clearConfigCache();
 * const freshConfig = resolveConfig(undefined, "./new-config.yml");
 * ```
 */
export function clearConfigCache(): void {
  cachedConfig = null;
  cachedConfigPath = null;
}

/**
 * Resolve config input to a ResolvedWritingConfig.
 *
 * Attempts to interpret the input string and return appropriate config.
 * Supports multiple input formats in order of precedence:
 * - undefined/null/empty: returns default "unknown" config
 * - File path: "./config.yml" or "/absolute/path.yml" -> loads full YAML config
 * - Media type: "blog", "book", "academic" -> loads preset for that media
 * - Preset name: "technical-book", "student-essay" -> loads composite preset
 * - JSON string: '{"media":"blog"}' -> extracts media type from JSON
 * - Falls back to "unknown" with a warning if unrecognized
 *
 * Results are cached by file path to avoid redundant I/O. Call {@link clearConfigCache}
 * to force a fresh reload.
 *
 * @param input - Config input (string, null, or undefined)
 * @param configPath - Optional file path to config file. When provided, overrides input
 *                     and loads the config file directly.
 * @returns Resolved writing config
 */
export function resolveConfig(
  input?: string | null,
  configPath?: string,
): ResolvedWritingConfig {
  // If config path is provided, use it directly
  if (configPath) {
    if (cachedConfig && cachedConfigPath === configPath) {
      return cachedConfig;
    }
    cachedConfig = loadConfig(configPath);
    cachedConfigPath = configPath;
    return cachedConfig;
  }

  // Default context
  if (input === undefined || input === null || input === '') {
    return loadFromMediaType('unknown');
  }

  // String input
  const inputStr = input.trim();

  // Check if it's a file path
  if (
    inputStr.endsWith('.yaml') ||
    inputStr.endsWith('.yml') ||
    inputStr.startsWith('./') ||
    inputStr.startsWith('/')
  ) {
    if (existsSync(inputStr)) {
      return loadConfig(inputStr);
    }
  }

  // Check if it's a media type
  if (isMediaType(inputStr)) {
    return loadFromMediaType(inputStr as MediaType);
  }

  // Check if it's JSON format with media type
  if (inputStr.startsWith('{')) {
    try {
      const parsed = JSON.parse(inputStr) as Record<string, unknown>;
      const media = parsed.media as string | undefined;
      if (media && isMediaType(media)) {
        return loadFromMediaType(media as MediaType);
      }
      // Invalid JSON format, try as preset
    } catch {
      // Invalid JSON, try as preset
    }
  }

  // Check if it's a preset name
  try {
    return loadPreset(inputStr);
  } catch {
    // Not a valid preset
  }

  // Unknown input, default to unknown media
  console.error(`Warning: Unknown config input "${inputStr}", using default`);
  return loadFromMediaType('unknown');
}

/** @deprecated Use resolveConfig instead */
export const resolveContext = resolveConfig;

/**
 * Serialize config for environment variable.
 * Re-export from config-loader for convenience.
 */
export { serializeConfigForEnv };

/**
 * Deserialize config from environment variable.
 * Re-export from config-loader for convenience.
 */
export { deserializeConfigFromEnv };

/**
 * Get all available presets.
 *
 * Returns a list of all built-in preset names that can be passed to
 * {@link resolveConfig}.
 *
 * @returns Array of preset names
 */
export function getPresets(): string[] {
  return [...getBuiltinPresets()];
}

/**
 * Get all available media types.
 *
 * Returns a list of all supported media types that can be passed to
 * {@link resolveConfig}.
 *
 * @returns Array of media type names
 */
export function getMediaTypes(): MediaType[] {
  return [...MEDIA_TYPES];
}

/**
 * Load writing config from a YAML file.
 *
 * Low-level function for loading config from a file path. Most consumers should use
 * {@link resolveConfig} instead, which automatically detects file paths and handles
 * caching.
 */
export { loadConfig };

/**
 * Load writing config from a preset name.
 *
 * Low-level function for loading a built-in preset by name. Most consumers should use
 * {@link resolveConfig} instead, which automatically detects preset names.
 */
export { loadPreset };

/**
 * Load writing config from a media type.
 *
 * Low-level function for loading config for a specific media type. Most consumers should use
 * {@link resolveConfig} instead, which automatically detects media types.
 */
export { loadFromMediaType };

/**
 * Get all available built-in presets.
 *
 * Low-level function that returns preset names. Most consumers should use
 * {@link getPresets} instead.
 */
export { getBuiltinPresets };
